flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/bundle/AbstractMapBundleOperator.java

```diff
@@ -0,0 +1,210 @@
+			BundleTrigger<IN> bundleTrigger) {
+		chainingStrategy = ChainingStrategy.ALWAYS;
+		this.bundle = new HashMap<>();
+		this.function = checkNotNull(function, "function is null");
+		this.bundleTrigger = checkNotNull(bundleTrigger, "bundleTrigger is null");
+	}
+
+	@Override
+	public void setup(
+			StreamTask<?, ?> containingTask,
+			StreamConfig config,
+			Output<StreamRecord<OUT>> output) {
+		super.setup(containingTask, config, output);
+		this.checkpointingLock = getContainingTask().getCheckpointLock();
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+		function.open(new ExecutionContextImpl(this, getRuntimeContext()));
+
+		this.numOfElements = 0;
+		this.collector = new StreamRecordCollector<>(output);
+
+		bundleTrigger.registerCallback(this);
+		// reset trigger
+		bundleTrigger.reset();
+		LOG.info("BundleOperator's trigger info: " + bundleTrigger.explain());
+
+		// counter metric to get the size of bundle
+		getRuntimeContext().getMetricGroup().gauge("bundleSize", (Gauge<Integer>) () -> numOfElements);
+		getRuntimeContext().getMetricGroup().gauge("bundleRatio", (Gauge<Double>) () -> {
+			int numOfKeys = bundle.size();
+			if (numOfKeys == 0) {
+				return 0.0;
+			} else {
+				return 1.0 * numOfElements / numOfKeys;
+			}
+		});
+	}
+
+	@Override
+	public void processElement(StreamRecord<IN> element) throws Exception {
+		while (isInFinishingBundle) {
+			checkpointingLock.wait();
+		}
+
+		// get the key and value for the map bundle
+		final IN input = element.getValue();
+		final K bundleKey = getKey(input);
+		final V bundleValue = bundle.get(bundleKey);
+
+		// get a new value after adding this element to bundle
+		final V newBundleValue = function.addInput(bundleValue, input);
+
+		// update to map bundle
+		bundle.put(bundleKey, newBundleValue);
+
+		numOfElements++;
+		bundleTrigger.onElement(input);
+	}
+
+	/**
+	 * Get the key for current processing element, which will be used as the map bundle's key.
+	 */
+	protected abstract K getKey(final IN input) throws Exception;
+
+	@Override
+	public void finishBundle() throws Exception {
+		while (isInFinishingBundle) {
+			checkpointingLock.wait();
+		}
+		isInFinishingBundle = true;
+		if (!bundle.isEmpty()) {
+			numOfElements = 0;
+			function.finishBundle(bundle, collector);
+			bundle.clear();
+		}
+		// reset trigger
+		bundleTrigger.reset();
+		checkpointingLock.notifyAll();
+		isInFinishingBundle = false;
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		while (isInFinishingBundle) {
+			checkpointingLock.wait();
+		}
+		finishBundle();
+		super.processWatermark(mark);
+	}
+
+	@Override
+	public void prepareSnapshotPreBarrier(long checkpointId) throws Exception {
+		while (isInFinishingBundle) {
+			checkpointingLock.wait();
+		}
+		finishBundle();
+	}
+
+	@Override
+	public void close() throws Exception {
+		while (isInFinishingBundle) {
+			checkpointingLock.wait();
+		}
+		try {
+			finishBundle();
```

Comment:
I remember Flink didn't allow data to be sent in `close`, only after introducing `endInput`?

Reply:
Ignore it. Sent data in `close` is OK.

Reply:
according to api doc of `StreamOperator::close()`:
> The method is expected to flush all remaining buffered data. Exceptions during this flushing of buffered should be propagated, in order to cause the operation to be recognized as failed, because the last data items are not processed properly

I think it's ok to emit data during close